// @workos-inc/node/lib/user-management/serializers/email-verification.serializer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.deserializeEmailVerificationEvent = exports.deserializeEmailVerification = void 0;
const deserializeEmailVerification = (emailVerification) => ({
    object: emailVerification.object,
    id: emailVerification.id,
    userId: emailVerification.user_id,
    email: emailVerification.email,
    expiresAt: emailVerification.expires_at,
    code: emailVerification.code,
    createdAt: emailVerification.created_at,
    updatedAt: emailVerification.updated_at,
});
exports.deserializeEmailVerification = deserializeEmailVerification;
const deserializeEmailVerificationEvent = (emailVerification) => ({
    object: emailVerification.object,
    id: emailVerification.id,
    userId: emailVerification.user_id,
    email: emailVerification.email,
    expiresAt: emailVerification.expires_at,
    createdAt: emailVerification.created_at,
    updatedAt: emailVerification.updated_at,
});
exports.deserializeEmailVerificationEvent = deserializeEmailVerificationEvent;
