// @workos-inc/node/lib/api-keys/api-keys.spec.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const jest_fetch_mock_1 = __importDefault(require("jest-fetch-mock"));
const test_utils_1 = require("../common/utils/test-utils");
const workos_1 = require("../workos");
const validate_api_key_json_1 = __importDefault(require("./fixtures/validate-api-key.json"));
describe('ApiKeys', () => {
    let workos;
    beforeAll(() => {
        workos = new workos_1.WorkOS('sk_test_Sz3IQjepeSWaI4cMS4ms4sMuU', {
            apiHostname: 'api.workos.test',
            clientId: 'proj_123',
        });
    });
    beforeEach(() => jest_fetch_mock_1.default.resetMocks());
    describe('validateApiKey', () => {
        it('sends a validate API key request', () => __awaiter(void 0, void 0, void 0, function* () {
            (0, test_utils_1.fetchOnce)(validate_api_key_json_1.default);
            const response = yield workos.apiKeys.validateApiKey({
                value: 'sk_123',
            });
            expect((0, test_utils_1.fetchURL)()).toContain('/api_keys/validations');
            expect(response).toEqual({
                apiKey: {
                    object: 'api_key',
                    id: 'api_key_01H5JQDV7R7ATEYZDEG0W5PRYS',
                    owner: {
                        type: 'organization',
                        id: 'org_01H5JQDV7R7ATEYZDEG0W5PRYS',
                    },
                    name: 'Test Api Key',
                    obfuscatedValue: 'sk_…PRYS',
                    lastUsedAt: null,
                    permissions: ['read', 'write'],
                    createdAt: '2023-07-18T02:07:19.911Z',
                    updatedAt: '2023-07-18T02:07:19.911Z',
                },
            });
        }));
        it('returns null if key is invalid', () => __awaiter(void 0, void 0, void 0, function* () {
            (0, test_utils_1.fetchOnce)({ api_key: null });
            const response = yield workos.apiKeys.validateApiKey({
                value: 'invalid',
            });
            expect((0, test_utils_1.fetchURL)()).toContain('/api_keys/validations');
            expect(response).toEqual({ apiKey: null });
        }));
    });
});
